### PR TITLE
[4.0] Migrate metadata manager listener to a proper service

### PR DESCRIPTION
**Pull request by mbabker**
### Summary of Changes

https://github.com/joomla/joomla-cms/pull/23433 was a bit hacky in creating the event listener for the session metadata manager.  This moves it to a more correct architecture solution.  Why does this matter?  It allows the event to be properly tested without having to load a boatload of unrelated code is part of it.  To accomplish this, the following changes are made:

- Create a new `LazyServiceEventListener`, this is a decorator for a service from the DI container which needs to be lazy loaded at runtime (whereas event subscribers/listeners generally have to be instantiated when registered to the dispatcher, see old PR for why this is relevant)
- Moves the event listener out from being a Closure in the Session service provider to its own class

### Testing Instructions

- Session metadata is still correctly processed, check database for this

### CODE REVIEW

**Comment by mbabker** [comment]
Nevermind, janky code it is.

**Comment by SharkyKZ** [comment]
Sir @mbabker could you re-open this if it's not too much trouble?

File: libraries/src/Event/LazyServiceEventListener.php [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	/**
+	 * Constructor.
+	 *
+	 * @param   string  $serviceId  The ID of the service from the container to be used
+	 * @param   string  $method     The method from the service to be called if necessary
+	 *                              (if left empty, the service must be a callable; i.e. have an `__invoke()` method on a class)
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function __construct(string $serviceId, string $method = '')
```
Comment (SharkyKZ) [inline libraries/src/Event/LazyServiceEventListener.php:51]: Add `@throws` to docblock?

File: libraries/src/Event/LazyServiceEventListener.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	/**
+	 * Load a service from the container to listen to an event.
+	 *
+	 * @param   EventInterface  $event  The event to process
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function __invoke(EventInterface $event)
```
Comment (SharkyKZ) [inline libraries/src/Event/LazyServiceEventListener.php:78]: Add `@throws` to docblock?

File: libraries/src/Event/LazyServiceEventListener.php (hunk 1 above)
Comment (mbabker) [reply libraries/src/Event/LazyServiceEventListener.php:51]: Added

File: libraries/src/Event/LazyServiceEventListener.php (hunk 2 above)
Comment (mbabker) [reply libraries/src/Event/LazyServiceEventListener.php:78]: Added

File: libraries/src/Event/LazyServiceEventListener.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	/**
+	 * Load a service from the container to listen to an event.
+	 *
+	 * @param   EventInterface  $event  The event to process
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 * @throws  \InvalidArgumentException if the constructor's $method parameter is empty when not executing a callable service
+	 * @throws  \RuntimeException if the container has not been injected, the service does not exist, or the constructor's $method does not exist in the service class
```
Comment (SharkyKZ) [inline libraries/src/Event/LazyServiceEventListener.php]: Drone please:

```
76 | WARNING | Line exceeds 150 characters; contains 163 characters
```

Comment (mbabker) [reply libraries/src/Event/LazyServiceEventListener.php]: It's 2019.  There is no place for a hard maximum line length rule in a code style enforcer.  Message changed.

**Comment by SharkyKZ** [comment]
I have tested this item :white_check_mark: successfully on bb62224d190b3e56097a61aa1d470c48eb40e2e6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23708">issues.joomla.org/tracker/joomla-cms/23708</a>.</sub>

**Comment by SharkyKZ** [comment]
Not related to this PR but noticed that in J4 on first visit `client_id` value in session table is `NULL`. Is this a bug?

**Comment by joeforjoomla** [comment]
@SharkyKZ this was a 'bug' i reported some months ago, but in the current Alpha 7 this seems to be fixed. On first visit the client is now '0' and not NULL

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on bb62224d190b3e56097a61aa1d470c48eb40e2e6<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23708">issues.joomla.org/tracker/joomla-cms/23708</a>.</sub>

**Comment by mbabker** [comment]
IIRC part of the intent with #23433 was to fix that bug in the first place.  Although looking again it looks like the wrong pseudo-constants are being checked [here](https://github.com/joomla/joomla-cms/blob/8882d2d556ea5f1f13865222de9914178840d493/libraries/src/Session/MetadataManager.php#L125-L138) and this needs correcting.

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/23708">issues.joomla.org/tracker/joomla-cms/23708</a>.</sub>

**Comment by wilsonge** [comment]
PR for the issue michael mentioned https://github.com/joomla/joomla-cms/pull/24120

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**